File: dart/merge_sort.dart

merge_sort(List<int> arr, int low, int high){
  if(low==high) return;

  int mid= ((low+high)/2).floor();

  merge_sort(arr, low, mid);
  merge_sort(arr, mid+1, high);
  merge(arr, low,mid , high);

}

merge(List<int> arr, int low, int mid, int high){
  List<int > temp=[];

  int left= low;
  int right= mid+1;
  print(arr);

  while(left<=mid&&right<=high){
    if(arr[left]<arr[right]){
      temp.add(arr[left]);
      left+=1;

    }else{
      temp.add(arr[right]);
      right+=1;
    }
  }

  while(left<=mid){
    temp.add(arr[left]);
      left+=1;
  }
  while(right<=high){
    temp.add(arr[right]);
      right+=1;
  }
    print(temp);

  for(int i=low; i<=high;i++)
  {
    arr[i]=temp[i-low];
    print("i : ${i}, high: ${high}, low:${low}");
  }

}
void main(){
  List<int> elemts=[19,54,86,10,8,43];
  List<int> elemts2=[5,4,3,2,1,0];
  merge_sort(elemts2,0, elemts2.length-1);
  print(elemts2);
}

File: sorting/merge_sort.py
def ms(arr, low , high):
    if (low==high): return

    mid= (low+high)//2

    ms(arr, low,mid)
    ms(arr, mid+1, high)
    merge(arr, low ,mid, high)



    

def merge(arr, low, mid, high):
    temp=[]
    left=low
    right=mid+1
    
    print(f"arr before merge: {arr}")
    while(left<=mid and right<=high):
        if arr[left]<arr[right]:
            temp.append(arr[left])
            left+=1
        else:
            temp.append(arr[right])
            right+=1

    while(left<=mid):
        temp.append(arr[left])
        left+=1

    while(right<=high):
        temp.append(arr[right])
        right+=1

    print(f"temp: {temp}")

    for i in range(low,high+1):
        print(f"i :{i} low:{low} high: {high}")
      
        arr[i]=temp[i-low]
    print(f"arr: {arr}")

# timecomplexity is O(logn)

   
    

def merge_sort(arr, n):
    ms(arr, 0, n-1)
    print(f"second smallest :{arr[1]}\n second largest{arr[n-2]}")
   # print(elements)

if __name__=="__main__":
    elements= [19,54,86,10,8,43]
   # elements= [5,4,3,2,1,0]
   # selection_sort(elements)
    #bubble_sort(elements)
    merge_sort(elements, len(elements))
    print(elements)
